test/envoye2e/stats/stats_xds_test.go

```diff
@@ -0,0 +1,233 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package stats
+
+import (
+	"fmt"
+	"strconv"
+	"testing"
+	"time"
+
+	dto "github.com/prometheus/client_model/go"
+
+	"istio.io/proxy/test/envoye2e/driver"
+	"istio.io/proxy/test/envoye2e/env"
+)
+
+const StatsClientHTTPListener = `
+name: client{{ .N }}
+traffic_direction: OUTBOUND
+address:
+  socket_address:
+    address: 127.0.0.1
+    port_value: {{ .Vars.ClientPort }}
+filter_chains:
+- filters:
+  - name: envoy.http_connection_manager
+    config:
+      codec_type: AUTO
+      stat_prefix: client{{ .N }}
+      http_filters:
+      - name: envoy.filters.http.wasm
+        config:
+          config:
+            vm_config:
+              runtime: "envoy.wasm.runtime.null"
+              code:
+                inline_string: "envoy.wasm.metadata_exchange"
+            configuration: "test"
+      - name: envoy.filters.http.wasm
+        config:
+          config:
+            root_id: "stats_outbound"
+            vm_config:
+              runtime: envoy.wasm.runtime.null
+              code:
+                inline_string: "envoy.wasm.stats"
+            configuration: |
+              { "debug": "false", max_peer_cache_size: 20, field_separator: ";.;" }
+      - name: envoy.router
+      route_config:
+        name: client
+        virtual_hosts:
+        - name: client
+          domains: ["*"]
+          routes:
+          - match: { prefix: / }
+            route:
+              cluster: server
+              timeout: 0s
+`
+
+const StatsServerHTTPListener = `
+name: server{{ .N }}
+traffic_direction: INBOUND
+address:
+  socket_address:
+    address: 127.0.0.1
+    port_value: {{ .Vars.ServerPort }}
+filter_chains:
+- filters:
+  - name: envoy.http_connection_manager
+    config:
+      codec_type: AUTO
+      stat_prefix: server{{ .N }}
+      http_filters:
+      - name: envoy.filters.http.wasm
+        config:
+          config:
+            vm_config:
+              runtime: "envoy.wasm.runtime.null"
+              code:
+                inline_string: "envoy.wasm.metadata_exchange"
```

Comment:
Can you try running without mx filter?

Reply:
Failed the same way on clang-asan without it:
```diff --git a/test/envoye2e/stats/stats_xds_test.go b/test/envoye2e/stats/stats_xds_test.go
index edb0097..387ba4a 100644
--- a/test/envoye2e/stats/stats_xds_test.go
+++ b/test/envoye2e/stats/stats_xds_test.go
@@ -40,14 +40,6 @@ filter_chains:
       codec_type: AUTO
       stat_prefix: client{{ .N }}
       http_filters:
-      - name: envoy.filters.http.wasm
-        config:
-          config:
-            vm_config:
-              runtime: "envoy.wasm.runtime.null"
-              code:
-                inline_string: "envoy.wasm.metadata_exchange"
-            configuration: "test"
       - name: envoy.filters.http.wasm
         config:
           config:
@@ -85,14 +77,6 @@ filter_chains:
       codec_type: AUTO
       stat_prefix: server{{ .N }}
       http_filters:
-      - name: envoy.filters.http.wasm
-        config:
-          config:
-            vm_config:
-              runtime: "envoy.wasm.runtime.null"
-              code:
-                inline_string: "envoy.wasm.metadata_exchange"
-            configuration: "test"
       - name: envoy.filters.http.wasm
         config:
           config:
```

Reply:
Huh, removing both SD and MX still shows the same issue. @jplevyak I think you are right, this is a general envoy issue, not due to Wasm. I have no idea why SD+MX passes but not without them.

Reply:
Trying master of envoy:
- fails with timeout in /stats/prometheus
- same huge number of ListenerImpl destroyed observed (over 8k)
- same huge number of drain ticks observed during the timeout

Reply:
I have a theory that the client proxy pool is what causes the server listeners to accumulate.

Reply:
Seems to happen with sd test too:
https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_proxy/2514/proxy-presubmit-tsan/2861
https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_proxy/2514/proxy-presubmit-asan/2863

Reply:
OK, updated both tests to 100ms xDS refresh rate and drain time of 4s. That should stabilize it.